src/Raven.Server/Documents/AI/Embeddings/EmbeddingsGenerator.cs

```diff
@@ -0,0 +1,908 @@
+
+                                using (var json = context.ReadObject(docJson, docId))
+                                {
+                                    documentsStorage.Put(context, docId, null, json);
+                                }
+
+                                string embeddingHash = AttachmentsStorageHelper.CalculateHash(embedding.Span);
+                                attachmentsStorage.PutAttachment(context, docId, valueHash, EmbeddingAttachmentContentType,
+                                    embeddingHash, null, new ReadOnlyMemoryStream<byte>(embedding));
+                                operations++;
+                            }
+                        }
+                        break;
+                    }
+                }
+            }
+            return operations; 
+        }
+        
+        private DynamicJsonValue CreateEmbeddingCacheDocumentJson(DateTime expireAt)
+        {
+            return new DynamicJsonValue
+            {
+                [Constants.Documents.Metadata.Key] = new DynamicJsonValue
+                {
+                    [Constants.Documents.Metadata.Collection] = Constants.Documents.Collections.EmbeddingsCacheCollection,
+                    [Constants.Documents.Metadata.Expires] = expireAt
+                }
+            };
+        }
+        
+
+        public override IReplayableCommandDto<DocumentsOperationContext, DocumentsTransaction, MergedTransactionCommand<DocumentsOperationContext, DocumentsTransaction>> ToDto(DocumentsOperationContext context)
+        {
+            return new Replay(batch);
+        }
+
+        public class Replay(List<object> work) : IReplayableCommandDto<DocumentsOperationContext, DocumentsTransaction, MergedTransactionCommand<DocumentsOperationContext, DocumentsTransaction>>
+        {
+            public MergedTransactionCommand<DocumentsOperationContext, DocumentsTransaction> ToCommand(DocumentsOperationContext context, DocumentDatabase database)
+            {
+                return new PutEmbeddingsIntoCacheCommand(work);
+            }
+        }
+    }
+    
+    public void HandleDatabaseRecordChange(DatabaseRecord record)
+    {
+        if (record is null)
+            return;
+
+        foreach (EmbeddingsGenerationConfiguration configuration in record.EmbeddingsGenerations)
+        {
+            var identifier = new EmbeddingsGenerationTaskIdentifier(configuration.Identifier);
+            if (configuration.Disabled)
+            {
+                if (_workers.TryRemove(identifier, out var toDispose))
+                {
+                    _ = toDispose.ShutdownAsync();
+                }
+
+                continue;
+            }
+            
+            if (_workers.TryGetValue(identifier, out var existing) is false)
+            {
+                _ = _workers.GetOrAdd(identifier, CreateAiWorker).RunAsync();
```

Comment:
something I learned recently is that GetOrAdd (and friends) might call valueFactory even when there will be no update, we need to switch to Lazy for the value

Reply:
The problem with using `Lazy` here is that you'll require a delegate allocation for every such call. 
Note that `HandleDatabaseRecordChange` is the only place that calls `GetOrAdd`

The actual spawning of the task is handled with (thread safe) `RunAsync`, so creating spurious instances via this is fine.